### ISSUE TITLE
data race in x509 check code

### ISSUE THREAD

**Issue by RuurdBeerstra** [openssl/openssl#7656]
Hi,

We use OpenSSL in a multi-threaded Windows program. Each login is handled by a thread, so we rely heavily on thread safety in OpenSSL. We use Intel Inspector to check for threading and memory issues, it found a few problems in OpenSSL 1.1.1 code. 

When a certificate is validated, 2 (or more) threads are in a race to fill and use the X509 certificate store.
Specifically, the code in v3_purp.c:486 writes to rfc3779_addr, v3_addr.c:1204 reads at the same time.
That is just one of 4 data races that Intel Inspector reports,
I've attached all four problem reports (as generated by Inspector).

Insights appreciated;
   Ruurd Beerstra.

[OpenSSLdatarace-2.txt](https://github.com/openssl/openssl/files/2595461/OpenSSLdatarace-2.txt)
[OpenSSLdatarace-1.txt](https://github.com/openssl/openssl/files/2595462/OpenSSLdatarace-1.txt)
[OpenSSLdatarace-4.txt](https://github.com/openssl/openssl/files/2595463/OpenSSLdatarace-4.txt)
[OpenSSLdatarace-3.txt](https://github.com/openssl/openssl/files/2595464/OpenSSLdatarace-3.txt)

**Comment by RuurdBeerstra** [comment]
Are you saying I should try that myself (locking the x509 store) and re-test with Inspector?
I'm not an OpenSSL maintainer, but I guess I could come up with a patch.
But I'd prefer if someone with knowledge about OpenSSL would create a patch, and I can re-test under Inspector to see if the data-races are gone.

**Comment by richsalz** [comment]
OpenSSL does not promise that all its datastructures are thread-safe, so yes you have to do your own locking before calling some API's.

**Comment by RuurdBeerstra** [comment]
I just read your post on OpenSSL and threads (https://www.openssl.org/blog/blog/2017/02/21/threads/).

I, for one, was under the impression that OpenSSL could safely be used in a multi-threaded environment.
We *did* have a (global) lock of our own around some of the API calls, in this case the SSL accept call. But that caused a complete block of our server when a client that was logging in misbehaved: The accept not only does the low-level TCP/IP accept, it also negotiates with the client (certificate checks).
So, we set a global lock, do the SSL accept, and a read on the socket will block because the client is not sending anything: All other threads are waiting for lock to be released and the whole program is stuck until the read times out (can be many minutes to hours, depending on circumstances, keepalives, etc.).
So, we removed the lock, the program is no longer stuck on a single misbehaving client, but now Intel Inspector (nice tool) sees data races in the OpenSSL x.509 code that look to me like they might cause real problems.
I saw that there is lock-management code in OpenSSL, and assumed that adding a few calls in the right places will fix this. Must be careful there not to introduce new problems when network problems cause delays when checking a certificate (might block all threads), but that should be doable, right?

I'd appreciate advice on what to do now.
If the statement in your above-mentioned post means: You cannot safely use OpenSSL in a multi-threaded program (for example because the various threads share the certificate store and that part of OpenSSL is not thread-safe but cannot be avoided) than we have a serious problem here.
If you say: OpenSSL can be safely used in a multi-threaded environment provided that the threads keep their objects nicely separated, then this issue can (I think) be classified as a problem in the OpenSSL code and should be fixed.
Because each of our threads is handling a single login, has it's own SSL context and shares nothing with other threads, which is how we interpreted "can safely be used".... 

Insights (very much) appreciated.
